### DISCUSSION TITLE
How to terminate app cleanly with SIGINT when using threads

### DISCUSSION THREAD

**Discussion by kajmaj87** [ArthurSonzogni/FTXUI#401] [Q&A] [answered]
I'm having trouble retaining the Ctrl+C functionality build in into this library when I add a separate thread to my program. I'm currently having something along these lines:

```
std::atomic<bool> running = true;
std::function<void()> shutdown_handler;
void signal_handler(int signal) {
    shutdown_handler();
}

... somewhere at the end of main() :

  auto screen = ScreenInteractive::FitComponent();
  screen.Loop(component_renderer);

  shutdown_handler = []() {
    running = false;
  };

  std::thread render_loop([&screen]() {
    signal(SIGINT, signal_handler);
    while(running){
      using namespace std::chrono_literals;
      std::this_thread::sleep_for(60ms);
      screen.PostEvent(Event::Custom); 
    }
  });

  render_loop.join();
}
```

After introducing the thread the application never ends completely after pressing Ctrl+C, the screen disappears, but the terminal hangs. I tried to handle this using my own signal_handler as seen above but that does not help. 
Is there a way to cleanly stop all the threads and cleanup the terminal correctly?

**Comment by ArthurSonzogni** [comment, marked as answer]
Hello,
Some questions about your example:
- Why are you starting the thread after the execution of the FTXUI loop? Why not before, so that they can run both in parallel.
- Why do you replace the signal handler from your thread instead of the main one?

**Reply by kajmaj87** [reply]
Hi, thanks for looking into this.
1. This was a mistake I made when creating this smaller example, in my real project there are two threads and they are both started before ftxui loop. I've changed this in smaller example also but it didn't help.
2. This is the first time I looked into signals in C++ and this was just a try to make it work, I tried all the possible combinations of replacing the signal handler in main, thread or both at once. 
I've updated the example to the best one I've found till now (this time I've included 3rd thread - in my main app this is the game loop):
```
  auto screen = ScreenInteractive::FitComponent();

  shutdown_handler = []() { running = false; };

  signal(SIGINT, signal_handler);

  std::thread render_loop([&screen]() {
    signal(SIGINT, signal_handler);
    while (running) {
      using namespace std::chrono_literals;
      std::this_thread::sleep_for(60ms);
      screen.PostEvent(Event::Custom);
    }
    screen.ExitLoopClosure()();
  });

  std::thread engine_loop([]() {
    signal(SIGINT, signal_handler);
    auto i = 1;
    while (running) {
      i++;
    }
    screen.ExitLoopClosure()();
  });

  screen.Loop(component_renderer);
  render_loop.join();
  engine_loop.join();

  return 0;
}
```
In this code above Ctrl+C closes the application correctly in around 9/10 times but hangs the terminal in the other case.

**Reply by kajmaj87** [reply]
I'm also wondering why this minimal example does not end cleanly:
```
  auto screen = ScreenInteractive::FitComponent();

  shutdown_handler = []() { running = false; };

  signal(SIGINT, signal_handler);
  screen.Loop(component_renderer);

  while (running){
  }

  return 0;
}
```
After I press Ctrl+C the drawn screen disappears but the application is still running. Does this mean that the FTXUI consumed the signal and my handler was never called? Is there any way to modify its signal handling to include my handler?
@ArthurSonzogni  Any thoughts on this?

**Reply by kajmaj87** [reply]
Ok, I've figured it out! I do not add any signal handlers at all but do this instead (relying on FTXUI internal handling of this):
```
  auto screen = ScreenInteractive::FitComponent();
  std::thread render_loop([&screen]() {
    while (running) {
       // some logic ...
    }
  });

  std::thread engine_loop([]() {
    while (running) {
      // some logic ...
    }
  });

  screen.Loop(component_renderer);
  running = false; // this will take care of finishing all my threads
  render_loop.join();
  engine_loop.join();

  return 0;
}

**Reply by ArthurSonzogni** [reply]
Yes the last one is the right way to handle this.
Internally FTXUI catch SIGINT, then it restores the terminal state and exit the loop.

**Reply by kajmaj87** [reply]
Thanks for your help! Great library btw, incredible what you've done here.

**Reply by ArthurSonzogni** [reply]
Thank you!
